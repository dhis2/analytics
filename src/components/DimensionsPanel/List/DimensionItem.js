import React, { Component } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import LockIcon from '@material-ui/icons/Lock'

import DimensionLabel from './DimensionLabel'
import RecommendedIcon from './RecommendedIcon'
import OptionsButton from './OptionsButton'
import DynamicDimensionIcon from '../../../assets/DynamicDimensionIcon'
import { getPredefinedDimensionProp } from '../../../modules/predefinedDimensions'
import { styles } from './styles/DimensionItem.style'

export class DimensionItem extends Component {
    state = { mouseOver: false }

    onOptionsClick = id => event => this.props.onOptionsClick(event, id)

    onMouseOver = () => {
        this.setState({ mouseOver: true })
    }

    onMouseExit = () => {
        this.setState({ mouseOver: false })
    }

    getDimensionIcon = () => {
        const Icon = getPredefinedDimensionProp(this.props.id, 'icon')
        return Icon ? (
            <Icon style={styles.fixedDimensionIcon} />
        ) : (
            <DynamicDimensionIcon style={styles.dynamicDimensionIcon} />
        )
    }

    getDimensionType = () => {
        const { id, name, isDeactivated } = this.props

        return (
            <span
                data-dimensionid={id}
                style={{
                    ...styles.text,
                    ...(isDeactivated ? styles.textDeactivated : {}),
                }}
            >
                {name}
            </span>
        )
    }

    render() {
        const {
            id,
            isDeactivated,
            isSelected,
            isRecommended,
            isLocked,
            onClick,
            onOptionsClick,
            innerRef,
            style,
            ...rest
        } = this.props

        const Icon = this.getDimensionIcon()
        const Label = this.getDimensionType()
        const itemStyle =
            isSelected && !isDeactivated
                ? { ...styles.item, ...styles.selected }
                : styles.item

        return (
            <li
                onMouseOver={this.onMouseOver}
                onMouseLeave={this.onMouseExit}
                ref={innerRef}
                style={Object.assign({}, itemStyle, style)}
                {...rest}
            >
                <DimensionLabel
                    id={id}
                    isDeactivated={isDeactivated}
                    onClick={onClick}
                >
                    <div style={styles.iconWrapper}>{Icon}</div>
                    <div style={styles.labelWrapper}>
                        {Label}
                        <RecommendedIcon isRecommended={isRecommended} />
                    </div>
                    {isLocked && (
                        <div style={styles.iconWrapper}>
                            <LockIcon style={styles.lockIcon} />
                        </div>
                    )}
                </DimensionLabel>
                {onOptionsClick ? (
                    <div style={styles.optionsWrapper}>
                        {this.state.mouseOver && !isDeactivated && !isLocked ? (
                            <OptionsButton
                                style={styles.optionsButton}
                                onClick={this.onOptionsClick(id)}
                            />
                        ) : null}
                    </div>
                ) : null}
            </li>
        )
    }
}

DimensionItem.propTypes = {
    id: PropTypes.string.isRequired,
    isSelected: PropTypes.bool.isRequired, // XXX
    name: PropTypes.string.isRequired,
    innerRef: PropTypes.func,
    isDeactivated: PropTypes.bool,
    isLocked: PropTypes.bool,
    isRecommended: PropTypes.bool,
    style: PropTypes.object,
    onClick: PropTypes.func,
    onOptionsClick: PropTypes.func,
}

DimensionItem.defaultProps = {
    isDeactivated: false,
    isRecommended: false,
    isSelected: false,
    isLocked: false,
    onClick: Function.prototype,
    innerRef: Function.prototype,
    style: {},
}

export default DimensionItem
