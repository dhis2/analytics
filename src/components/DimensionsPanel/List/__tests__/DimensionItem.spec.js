import React from 'react'
import { DimensionItem } from '../DimensionItem'
import { shallow } from 'enzyme'

describe('DimensionItem', () => {
    let props
    let shallowItem
    const dimensionItem = () => {
        if (!shallowItem) {
            shallowItem = shallow(<DimensionItem {...props} />)
        }
        return shallowItem
    }

    beforeEach(() => {
        props = {
            id: 'pe',
            name: 'Period',
            isDeactivated: false,
            isSelected: false,
            isRecommended: false,
            isLocked: false,
        }

        shallowItem = undefined
    })

    it('matches the snapshot', () => {
        expect(dimensionItem()).toMatchSnapshot()
    })

    it('matches the snapshot with recommended', () => {
        props.isRecommended = true
        expect(dimensionItem()).toMatchSnapshot()
    })

    it('matches the snapshot with selected', () => {
        props.isSelected = true
        expect(dimensionItem()).toMatchSnapshot()
    })

    it('matches the snapshot with locked', () => {
        props.isLocked = true
        expect(dimensionItem()).toMatchSnapshot()
    })

    it('matches the snapshot with onOptionsClick', () => {
        props.onOptionsClick = jest.fn()
        expect(dimensionItem()).toMatchSnapshot()
    })
})
