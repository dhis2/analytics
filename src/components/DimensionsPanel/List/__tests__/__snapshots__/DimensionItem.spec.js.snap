// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DimensionItem matches the snapshot 1`] = `
<li
  name="Period"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": "2px",
      "display": "flex",
      "marginBottom": 3,
      "marginTop": 3,
      "minHeight": "24px",
    }
  }
>
  <DimensionLabel
    id="pe"
    isDeactivated={false}
    onClick={[Function]}
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <PeriodIcon
        style={
          Object {
            "paddingBottom": "2px",
            "paddingLeft": "6px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "maxWidth": "174px",
          "padding": "2px 0",
        }
      }
    >
      <span
        data-dimensionid="pe"
        style={
          Object {
            "color": "#000000",
            "cursor": "pointer",
            "fontSize": "14px",
            "userSelect": "none",
            "wordBreak": "break-word",
          }
        }
      >
        Period
      </span>
      <RecommendedIcon
        isRecommended={false}
      />
    </div>
  </DimensionLabel>
</li>
`;

exports[`DimensionItem matches the snapshot with locked 1`] = `
<li
  name="Period"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": "2px",
      "display": "flex",
      "marginBottom": 3,
      "marginTop": 3,
      "minHeight": "24px",
    }
  }
>
  <DimensionLabel
    id="pe"
    isDeactivated={false}
    onClick={[Function]}
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <PeriodIcon
        style={
          Object {
            "paddingBottom": "2px",
            "paddingLeft": "6px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "maxWidth": "174px",
          "padding": "2px 0",
        }
      }
    >
      <span
        data-dimensionid="pe"
        style={
          Object {
            "color": "#000000",
            "cursor": "pointer",
            "fontSize": "14px",
            "userSelect": "none",
            "wordBreak": "break-word",
          }
        }
      >
        Period
      </span>
      <RecommendedIcon
        isRecommended={false}
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <pure(LockIcon)
        style={
          Object {
            "fontSize": "14px",
          }
        }
      />
    </div>
  </DimensionLabel>
</li>
`;

exports[`DimensionItem matches the snapshot with onOptionsClick 1`] = `
<li
  name="Period"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": "2px",
      "display": "flex",
      "marginBottom": 3,
      "marginTop": 3,
      "minHeight": "24px",
    }
  }
>
  <DimensionLabel
    id="pe"
    isDeactivated={false}
    onClick={[Function]}
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <PeriodIcon
        style={
          Object {
            "paddingBottom": "2px",
            "paddingLeft": "6px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "maxWidth": "174px",
          "padding": "2px 0",
        }
      }
    >
      <span
        data-dimensionid="pe"
        style={
          Object {
            "color": "#000000",
            "cursor": "pointer",
            "fontSize": "14px",
            "userSelect": "none",
            "wordBreak": "break-word",
          }
        }
      >
        Period
      </span>
      <RecommendedIcon
        isRecommended={false}
      />
    </div>
  </DimensionLabel>
  <div
    style={
      Object {
        "alignSelf": "start",
        "height": "24px",
        "left": "5px",
        "position": "relative",
      }
    }
  />
</li>
`;

exports[`DimensionItem matches the snapshot with recommended 1`] = `
<li
  name="Period"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": "2px",
      "display": "flex",
      "marginBottom": 3,
      "marginTop": 3,
      "minHeight": "24px",
    }
  }
>
  <DimensionLabel
    id="pe"
    isDeactivated={false}
    onClick={[Function]}
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <PeriodIcon
        style={
          Object {
            "paddingBottom": "2px",
            "paddingLeft": "6px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "maxWidth": "174px",
          "padding": "2px 0",
        }
      }
    >
      <span
        data-dimensionid="pe"
        style={
          Object {
            "color": "#000000",
            "cursor": "pointer",
            "fontSize": "14px",
            "userSelect": "none",
            "wordBreak": "break-word",
          }
        }
      >
        Period
      </span>
      <RecommendedIcon
        isRecommended={true}
      />
    </div>
  </DimensionLabel>
</li>
`;

exports[`DimensionItem matches the snapshot with selected 1`] = `
<li
  name="Period"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#D9ECEB",
      "borderRadius": "2px",
      "display": "flex",
      "fontWeight": 500,
      "marginBottom": 3,
      "marginTop": 3,
      "minHeight": "24px",
    }
  }
>
  <DimensionLabel
    id="pe"
    isDeactivated={false}
    onClick={[Function]}
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "3px 8px 0 8px",
        }
      }
    >
      <PeriodIcon
        style={
          Object {
            "paddingBottom": "2px",
            "paddingLeft": "6px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "maxWidth": "174px",
          "padding": "2px 0",
        }
      }
    >
      <span
        data-dimensionid="pe"
        style={
          Object {
            "color": "#000000",
            "cursor": "pointer",
            "fontSize": "14px",
            "userSelect": "none",
            "wordBreak": "break-word",
          }
        }
      >
        Period
      </span>
      <RecommendedIcon
        isRecommended={false}
      />
    </div>
  </DimensionLabel>
</li>
`;
